### DISCUSSION TITLE
Is there any auth log for failed logins?

### DISCUSSION THREAD

**Discussion by netw0rk-noob** [healthchecks/healthchecks#1061] [Q&A] [answered]
The question is the one in the title, basically. I would like to setup fail2ban for the WebUI login but wasnt able to find any log containing (failed) login attempts.

**Comment by cuu508** [comment]
No, there is no auth log for failed logins.

There is however built-in rate-limiting:

* a single IP address is limited to 20 login attempts per hour
* a single email address is limited to 20 login attempts per hour, when logging in via the "email me a link" method
* a single email address is limited to 20 login attempts per day, when logging in via the email/password method

The rate-limiting is implemented here: https://github.com/healthchecks/healthchecks/blob/13217af304203fb3e842d4ad2c7cf317f129e1e7/hc/api/models.py#L1218

**Reply by netw0rk-noob** [reply]
That is indeed good to know, thank you. Is that feature documented anywhere (besides the source code)? (I'm asking because at least I didnt read anything about it in the selfhosted docs section, but I may have missed it.)

Are there any plans to make these limits configurable? Or (even better imho) to implement an auth logging mechanism on which de-facto standard tools like fail2ban could act upon?

**Reply by cuu508** [reply, marked as answer]
No, the rate limits are not documented, and I have no current plans to make them configurable.

Perhaps of interest to you, Healthchecks supports external authentication: https://healthchecks.io/docs/self_hosted_configuration/#REMOTE_USER_HEADER

**Reply by netw0rk-noob** [reply]
Thanks for the clarification.

**Reply by netw0rk-noob** [reply]
I tried setting up http basic auth with nginx on the same machine:
from my `local_settings.py` 
```
[…]
REMOTE_USER_HEADER = "HTTP_AUTHORIZATION"
```
from my `nginx` site config:
```
server {
[…]
        auth_basic                      "Healthchecks login";
        auth_basic_user_file            /etc/nginx/.htpasswd;
[…]
        location / {
                […]
                proxy_pass_header Authorization;
                […]
        }
        location /ping  {
                auth_basic off;
                proxy_pass_header Authorization;
                […]
        }
}
```
The healthchecks user with which I can login perfectly fine without above changes to my configs is set up with `htpasswd -c /etc/nginx/.htpasswd myuser@healthchecks.tld`.

The basic auth on nginx's part does work - Im getting asked for credentials and the page is delivered if I enter them correctly. What doesnt seem to work is this part from [the docs you linked](https://healthchecks.io/docs/self_hosted_configuration/#REMOTE_USER_HEADER):
> - in views that require authentication, look up the request header specified in REMOTE_USER_HEADER
> - assume the header contains the user's email address
> - **automatically log in the user with a matching email address**
> - **automatically create a user account if it does not exist**
> - **disable the default authentication methods (login link to email, password)**

After the basic auth I'm still redirected to `https://healthchecks.tld/accounts/login/`.
If I try to login there (again) with my correct credentials, the page is simply reloaded, without any hint to what happened (no "wrong Password/E-Mail" or anything alike). The only difference is that while reloading the login page, the _referrer_ to my project (`?next=/projects/some-long-uuid/checks/`) is appended to the URL, but even after multiple *correct* login attempts I'm not redirected there.

Now the obvious question ist: What am I doing wrong here?

**Reply by cuu508** [reply]
Two notes:

* The HTTP header nominated in `REMOTE_USER_HEADER` should contain the user's email address. If you use HTTP Basic authorization, my understanding is the Authorization header would contain `base64(username + ":" + password)`, not the email address.
* The HTTP header nominated in `REMOTE_USER_HEADER` should be present on each and every HTTP request that requires authenticated user.

**Reply by netw0rk-noob** [reply]
> - The HTTP header nominated in REMOTE_USER_HEADER should contain the user's email address. If you use HTTP Basic authorization, my understanding is the Authorization header would contain base64(username + ":" + password), not the email address.

Indeed, didnt realize that one. I've changed `proxy_pass_header Authorization;` in my nginx site config to `proxy_set_header X-Forwarded-User $remote_user;`. As I've registered the healthchecks email address as a user for http basic auth, this header should contain the email address in question, according to the [nginx docs](https://nginx.org/en/docs/http/ngx_http_core_module.html#var_remote_user). In my `local_settings.py` I changed `REMOTE_USER_HEADER = "HTTP_AUTHORIZATION"` to `REMOTE_USER_HEADER = "HTTP_X_FORWARDED_USER"` accordingly.

After these changes I'm still presented with the healthchecks login page after going through basic auth, but now I can login through the healthchecks interface successfully. While still not exactly the bahaviour the [REMOTE_USER_HEADER docs](https://healthchecks.io/docs/self_hosted_configuration/#REMOTE_USER_HEADER) describe, thats a step forward at least.

> - The HTTP header nominated in REMOTE_USER_HEADER should be present on each and every HTTP request that requires authenticated user.

I'm not sure I get the point youre making here. Does this refer to the fact that I've set nginx to also pass the `authorization` header in the `/ping` location? I've removed that directive, because it doesnt make sense imho, as the ping location doesnt need the request to be from an authenticated user. If that wasnt the point, I'd be very grateful if you could elaborate on that.

**Reply by netw0rk-noob** [reply]
@cuu508 Have you got any other hints for me? I'm not getting it, it seems.

**Reply by cuu508** [reply]
The

```
proxy_set_header X-Forwarded-User $remote_user;
```

in nginx.conf and 

```
REMOTE_USER_HEADER = "HTTP_X_FORWARDED_USER"
```

in local_settings.py look good to me.

What I think is happening is Healthchecks is still using the default authentication backend, not the custom one which inspects request headers. If you [take a look at settings.py](https://github.com/healthchecks/healthchecks/blob/71f92d19fa8fab45ba8c8e49516eda79611f9061/hc/settings.py#L95), the custom authentication backend gets activated conditionally if REMOTE_USER_HEADER is set. This works fine if you configure Healthchecks via env variables. In your case I think the problem is, at the time this code runs, REMOTE_USER_HEADER is not set yet. That happens later, at the very end of settings.py  runs  `from .local_settings import *` which then redefines REMOTE_USER_HEADER.

In `local_settings.py` try:

```
REMOTE_USER_HEADER = "HTTP_X_FORWARDED_USER"
AUTHENTICATION_BACKENDS = ["hc.accounts.backends.CustomHeaderBackend"]
```

FWIW I did a quick experiment. I installed nginx, set up a .htpasswd file the same way as you did. My nginx configuration:

```
server {
    listen 80 default_server;

    auth_basic                      "Healthchecks login";
    auth_basic_user_file            /etc/nginx/.htpasswd;

    location / {
        proxy_set_header Host $http_host;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_set_header X-Forwarded-User $remote_user;
        proxy_pass http://localhost:8000;
    }
}
```

And in local_settings.py among other things I have

```
REMOTE_USER_HEADER = "HTTP_X_FORWARDED_USER"
AUTHENTICATION_BACKENDS = ["hc.accounts.backends.CustomHeaderBackend"]
```

With this in place, an user account got created and I was logged in on the first request. I didn't see the login page.

**Reply by netw0rk-noob** [reply]
Thats it, indeed. When specifying the authentication backend manually (which I didnt know I could), it does work as documented.
Thank you very much for your time and effort!

**Reply by cuu508** [reply]
Happy to hear you got it working!

I added a note in docs about the AUTHENTICATION_BACKENDS gotcha: https://github.com/healthchecks/healthchecks/commit/0b7b77c8d740d171ed2ea34baf00119e03c8b54e
